### DISCUSSION TITLE
Custom on chain Message verification

### DISCUSSION THREAD

**Discussion by AntonAndell** [informalsystems/ibc-rs#467] [General]
Looking to deploy a light client in wasmd but do not want to go through the process of supplying a official light client. Would it be possible to modify this library to allow for custom verification of IBC messages? Then use the modified library as usual for IBC modules.

**Comment by plafer** [comment]
As currently designed, you can write you own light client in your own crate; no need to upstream it to ibc-rs (actually we don't *want* more light clients in the crate 🙂). Just implement [the `ClientState` trait](https://github.com/cosmos/ibc-rs/blob/9713ec8c3ca61a27653ab8cf3435cf0dfea3dcc9/crates/ibc/src/core/ics02_client/client_state.rs#L32), and on your host, give it to ibc-rs through [these methods](https://github.com/cosmos/ibc-rs/blob/9713ec8c3ca61a27653ab8cf3435cf0dfea3dcc9/crates/ibc/src/core/context.rs#L242-L245). Let me know if that's not clear; we're planning on making our docs better soon.

Is this what you meant by "official"?

**Reply by AntonAndell** [reply]
I am still quite new to to cosmos ecosystem, I might be misunderstanding something. 
In short i want to deploy a smart contract light client on a IBC enabled sdk chain without a fork. When i meant official i was probably referring to the clients defined in IBC-go.
I might be confusing on the distinction between IBC-go and IBC-rs.
However, what you describe sounds like the exact thing i am looking for if it requires no fork of the host chain.

**Reply by plafer** [reply]
> I am still quite new to to cosmos ecosystem

Welcome!

> i want to deploy a smart contract light client on a IBC enabled sdk chain without a fork

How you do this depends on the chain. Is it a cosmwasm-based chain by any chance? 

ibc-rs is probably not the right library for your purposes, but we can redirect you to the right project/library.

**Reply by AntonAndell** [reply]
Thanks!
Yeah a cosmwasm chain.
But then i guess what i am looking for is a implementation for IBC in Cosmwasm. But i'm guessing that does not really exist since the would be using the built in IBC module.

**Reply by plafer** [reply]
cosmwasm chains are currently built on top of the Cosmos SDK (which uses ibc-go), where cosmwasm is an SDK module. I don't know much more about it; you should probably go ask the same question on the chain's repo (probably?).
